controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +70,211 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
 	policy := &policyv1beta1.OperatorPolicy{}
 
-	err := r.Get(ctx, req.NamespacedName, policy)
+	watcher := depclient.ObjectIdentifier{
+		Group:     operatorv1.GroupVersion.Group,
+		Version:   operatorv1.GroupVersion.Version,
+		Kind:      "OperatorPolicy",
+		Namespace: req.Namespace,
+		Name:      req.Name,
+	}
+
+	// Start query batch for caching and watching related objects
+	err := r.DynamicWatcher.StartQueryBatch(watcher)
+	if err != nil {
+		panic(err)
+	}
+
+	// Get the applied OperatorPolicy
+	err = r.Get(ctx, req.NamespacedName, policy)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			OpLog.Info("Operator policy could not be found")
 
+			err := r.DynamicWatcher.EndQueryBatch(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error ending the query batch in progress. Ignoring the failure.")
+			}
+
+			err = r.DynamicWatcher.RemoveWatcher(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error updating dependency watcher. Ignoring the failure.")
+			}
+
 			return reconcile.Result{}, nil
 		}
+
+		OpLog.Error(err, "Failed to get operator policy")
+
+		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, nil
-}
+	// handle the policy
+	OpLog.Info("Reconciling OperatorPolicy", "policy", policy.Name)
 
-// PeriodicallyExecOperatorPolicies loops through all operatorpolicies in the target namespace and triggers
-// template handling for each one. This function drives all the work the operator policy controller does.
-func (r *OperatorPolicyReconciler) PeriodicallyExecOperatorPolicies(freq uint, elected <-chan struct{},
-) {
-	OpLog.Info("Waiting for leader election before periodically evaluating operator policies")
-	<-elected
+	// Check if specified Subscription exists
+	cachedItem, err := r.DynamicWatcher.Get(watcher, subscriptiongvk, policy.Spec.Subscription.Namespace,
+		policy.Spec.Subscription.SubscriptionSpec.Package)
+	if err != nil {
+		panic(err)
+	}
 
-	exiting := false
-	for !exiting {
-		start := time.Now()
+	subExists := cachedItem != nil
 
-		var skipLoop bool
-		if !skipLoop {
-			policiesList := policyv1beta1.OperatorPolicyList{}
+	// Check if any OperatorGroups exist in the namespace
+	ogNamespace := policy.Spec.Subscription.Namespace
+	if policy.Spec.OperatorGroup != nil {
+		ogNamespace = policy.Spec.OperatorGroup.Namespace
+	}
 
-			err := r.List(context.TODO(), &policiesList)
-			if err != nil {
-				OpLog.Error(err, "Failed to list the OperatorPolicy objects to evaluate")
-			} else {
-				for i := range policiesList.Items {
-					policy := policiesList.Items[i]
-					if !r.shouldEvaluatePolicy(&policy) {
-						continue
-					}
-
-					// handle policy
-					err := r.handleSinglePolicy(&policy)
-					if err != nil {
-						OpLog.Error(err, "Error while evaluating operator policy")
-					}
-				}
-			}
-		}
+	cachedList, err := r.DynamicWatcher.List(watcher, operatorgroupgvk, ogNamespace, nil)
```

Comment:
Is it common to have multiple operator groups in the same namespace? I'm wondering if this will cause a lot of extra reconciles since you are using the dynamic watch library.

Reply:
I wouldn't say it's common but it is possible to have multiple operator groups. However, having multiple operator groups in the same ns will transition any CSV within that ns to a failure state with the reason `TooManyOperatorGroups`. Using a `List` guards against the case where there is a preexisting operator group on the cluster (either created manually or by another operator policy) and you want to apply another operator policy that has a different operator group spec in the same ns. Since we don't know the name of other operator groups except the one we're currently reconciling, we just check that resource type within the ns.